v2/cmd/driver/main.go

```diff
@@ -0,0 +1,176 @@
+	if *driverType == "" {
+		return fmt.Errorf("argument --%s must be specified", driverTypeArg)
+	}
+	// validation responsibility lives in driver itself, so we do not validate all other args
+	return nil
+}
+
+func drive() error {
+	ctx := context.Background()
+	componentSpec := &pipelinespec.ComponentSpec{}
+	if err := jsonpb.UnmarshalString(*componentSpecJson, componentSpec); err != nil {
+		return fmt.Errorf("failed to unmarshal component spec, error: %w, job: %v", err, componentSpecJson)
+	}
+	taskSpec := &pipelinespec.PipelineTaskSpec{}
+	if err := jsonpb.UnmarshalString(*taskSpecJson, taskSpec); err != nil {
+		return fmt.Errorf("failed to unmarshal task spec, error: %w, task: %v", err, taskSpecJson)
+	}
+	runtimeConfig := &pipelinespec.PipelineJob_RuntimeConfig{}
+	if err := jsonpb.UnmarshalString(*runtimeConfigJson, runtimeConfig); err != nil {
+		return fmt.Errorf("failed to unmarshal runtime config, error: %w, runtimeConfig: %v", err, runtimeConfigJson)
+	}
+	client, err := newMlmdClient()
+	if err != nil {
+		return err
+	}
+	options := driver.Options{
+		PipelineName:   *pipelineName,
+		RunID:          *runID,
+		Component:      componentSpec,
+		Task:           taskSpec,
+		DAGExecutionID: *dagExecutionID,
+		DAGContextID:   *dagContextID,
+	}
+	var execution *driver.Execution
+	switch *driverType {
+	case "ROOT_DAG":
+		options.RuntimeConfig = runtimeConfig
+		execution, err = driver.RootDAG(ctx, options, client)
+	case "CONTAINER":
+		execution, err = driver.Container(ctx, options, client)
+		if err != nil {
+
+		}
+	default:
+		return fmt.Errorf("unknown driverType %s", *driverType)
+	}
+	if err != nil {
+		return err
+	}
+	if execution.ID != 0 {
+		err := writeFile(*executionIDPath, []byte(fmt.Sprint(execution.ID)))
+		if err != nil {
+			return fmt.Errorf("failed to write execution ID to file: %w", err)
+		}
+	}
+	if execution.Context != 0 {
+		err = writeFile(*contextIDPath, []byte(fmt.Sprint(execution.Context)))
+		if err != nil {
+			return fmt.Errorf("failed to write context ID to file: %w", err)
+		}
+	}
+	if execution.ExecutorInput != nil {
+		marshaler := jsonpb.Marshaler{}
+		executorInputJson, err := marshaler.MarshalToString(execution.ExecutorInput)
+		if err != nil {
+			return fmt.Errorf("failed to marshal ExecutorInput to JSON: %w", err)
+		}
+		err = writeFile(*executorInputPath, []byte(executorInputJson))
+		if err != nil {
+			return fmt.Errorf("failed to write ExecutorInput to file: %w", err)
+		}
+	}
+	return nil
+}
+
+func writeFile(path string, data []byte) error {
+	if path == "" {
+		return fmt.Errorf("path is not specified")
+	}
+	err := os.MkdirAll(filepath.Dir(path), 0o755)
+	if err == nil { // no error
+		err = ioutil.WriteFile(path, data, 0o644)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to write to %s: %w", path, err)
+	}
+	return nil
+}
+
+func newMlmdClient() (*metadata.Client, error) {
+	mlmdConfig := metadata.DefaultConfig()
+	if *mlmdServerAddress != "" && *mlmdServerPort != "" {
```

Comment:
Do we need to throw error when *mlmdServerAddress == "" || *mlmdServerPort == ""?

Reply:
I don't think that's necessary, because there's always a non-empty default config.